### PR TITLE
Add support for `AssemblyLevel::FULL` for DG on `NCMesh`.

### PR DESCRIPTION
**Pull request by YohannDudouit**
This Pull Request adds support for sparse-matrix assembly with device support for DG on `NCMesh`, both serial and parallel sparse matrix assembly are supported.

This new feature can be tested on `ex9` and `ex9p` using the `-fa` option and using a nonconforming mesh.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4056](https://github.com/mfem/mfem/pull/4056) | @YohannDudouit | @pazner | @rw-anderson + @artv3 | 1/11/24 | 4/3/24 | ⌛due 4/3/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@artv3 and @rw-anderson -- can you please review when you get a chance?

**Comment by tzanio** [comment]
ping: @artv3, @rw-anderson

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

**Review by artv3 — APPROVED** [review]

**Review by acfisher — APPROVED** [review]
Look good.  It was great that the skeleton was already in place so the implementation just needed to be filled in.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
@pazner, I started the testing for this to help expedite it for `mfem-4.7` but am happy to hold on merging if you want to take a closer look

File: fem/prestriction.cpp [hunk 1]
```diff
@@ -858,11 +882,94 @@ void ParNCL2FaceRestriction::FillI(SparseMatrix &mat,
...
+      const int iF = fdof%nface_dofs;
+      const int iE1 = d_indices1[f*nface_dofs+iF];
+      if (iE1 < Ndofs)
+      {
+         const int offset1 = AddNnz(iE1,I,nface_dofs);
+         for (int jF = 0; jF < nface_dofs; jF++)
+         {
+            const int jE2 = d_indices2[f*nface_dofs+jF];
+            J[offset1+jF] = jE2;
+            double val2 = 0.0;
```
Comment (v-dobrev) [inline fem/prestriction.cpp:920]: There are several places, like this, where `double` needs to be replaced by `real_t`.

Comment (tzanio) [reply fem/prestriction.cpp:920]: @YohannDudouit, can you make this ☝️ change?

Comment (YohannDudouit) [inline fem/prestriction.cpp:920]:
```suggestion
            real_t val2 = 0.0;
```

File: fem/prestriction.cpp [hunk 2]
```diff
@@ -858,11 +882,94 @@ void ParNCL2FaceRestriction::FillI(SparseMatrix &mat,
...
+      }
+      const int iE2 = d_indices2[f*nface_dofs+iF];
+      if (iE2 < Ndofs)
+      {
+         const int offset2 = AddNnz(iE2,I,nface_dofs);
+         for (int jF = 0; jF < nface_dofs; jF++)
+         {
+            const int jE1 = d_indices1[f*nface_dofs+jF];
+            J[offset2+jF] = jE1;
+            double val1 = 0.0;
```
Comment (YohannDudouit) [inline fem/prestriction.cpp:950]:
```suggestion
            real_t val1 = 0.0;
```

File: fem/restriction.cpp [hunk 3]
```diff
@@ -2046,22 +2046,205 @@ void NCL2FaceRestriction::AddMultTransposeInPlace(Vector& x, Vector& y) const
...
+      const int offset1 = AddNnz(iE1,I,nface_dofs);
+      const int offset2 = AddNnz(iE2,I,nface_dofs);
+      for (int jF = 0; jF < nface_dofs; jF++)
+      {
+         const int jE1 = d_indices1[f*nface_dofs+jF];
+         const int jE2 = d_indices2[f*nface_dofs+jF];
+         J[offset2+jF] = jE1;
+         J[offset1+jF] = jE2;
+         double val1 = 0.0;
+         double val2 = 0.0;
```
Comment (YohannDudouit) [inline fem/restriction.cpp:2095]:
```suggestion
         real_t val1 = 0.0;
         real_t val2 = 0.0;
```

File: fem/restriction.cpp [hunk 4]
```diff
@@ -2046,22 +2046,205 @@ void NCL2FaceRestriction::AddMultTransposeInPlace(Vector& x, Vector& y) const
...
+         const int interp_index = conf.index;
+         const int e1 = d_indices1[f*nface_dofs]/nelem_dofs;
+         const int e2 = d_indices2[f*nface_dofs]/nelem_dofs;
+         for (int j = 0; j < nface_dofs; j++)
+         {
+            const int jB1 = d_indices1[f*nface_dofs+j]%nelem_dofs;
+            for (int i = 0; i < nface_dofs; i++)
+            {
+               const int iB1 = d_indices1[f*nface_dofs+i]%nelem_dofs;
+               double val = 0.0;
```
Comment (YohannDudouit) [inline fem/restriction.cpp:2154]:
```suggestion
               real_t val = 0.0;
```

File: fem/restriction.cpp [hunk 5]
```diff
@@ -2046,22 +2046,205 @@ void NCL2FaceRestriction::AddMultTransposeInPlace(Vector& x, Vector& y) const
...
+         }
+         if (e2 < NE)
+         {
+            for (int j = 0; j < nface_dofs; j++)
+            {
+               const int jB2 = d_indices2[f*nface_dofs+j]%nelem_dofs;
+               for (int i = 0; i < nface_dofs; i++)
+               {
+                  const int iB2 = d_indices2[f*nface_dofs+i]%nelem_dofs;
+                  double val = 0.0;
```
Comment (YohannDudouit) [inline fem/restriction.cpp:2182]:
```suggestion
                  real_t val = 0.0;
```

File: fem/restriction.cpp [hunk 6]
```diff
@@ -2046,22 +2046,205 @@ void NCL2FaceRestriction::AddMultTransposeInPlace(Vector& x, Vector& y) const
...
+         const int master_side = conf.master_side;
+         const int interp_index = conf.index;
+         const int e = d_indices[f*nface_dofs]/nelem_dofs;
+         for (int j = 0; j < nface_dofs; j++)
+         {
+            const int jE = d_indices[f*nface_dofs+j]%nelem_dofs;
+            for (int i = 0; i < nface_dofs; i++)
+            {
+               const int iE = d_indices[f*nface_dofs+i]%nelem_dofs;
+               double val = 0.0;
```
Comment (YohannDudouit) [inline fem/restriction.cpp:2226]:
```suggestion
               real_t val = 0.0;
```

File: fem/prestriction.cpp (hunk 1 above)
Comment (YohannDudouit) [reply fem/prestriction.cpp]: done.

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**